torch/csrc/jit/codegen/cuda/graph_fuser.cpp

```diff
@@ -914,19 +916,134 @@ void PeepholeOptimizeShapeExpressions(Block* block) {
+    // We only check inputs of the fusion group and expect NNC to infer
+    // intermediates and outputs shapes
+    if (!input->type()->cast<TensorType>()) {
+      continue;
+    }
+
+    // fusion outputs are already guarded
+    if (input->node()->kind() == prim::Constant) {
+      // input->node()->kind() == prim::FusionGroup ||
+      // input->node()->kind() == prim::CudaFusionGroup) {
+      continue;
+    }
+    inputs_to_check.push_back(input);
+    guard_types.push_back(input->type());
+  }
+  if (!inputs_to_check.size()) {
+    return;
+  }
+
+  // Add prim::CudaFusionGuard node
+  //
+  // CudaFusionGuard nodes  look like the following:
+  //   %out1 : Float(2, 3), %out2 : Int(10, 30), %types_match : bool =
+  //   prim::CudaFusionGuard(%inp1 : Tensor, %inp2 : Tensor)
+  //
+  // They have N inputs whose types we are going to check and N+1 outputs. The
+  // first N outputs specify expected types and N+1-th output holds the result
+  // of the check (bool).
+  //
+  // See [ Note -- type guard logic in CudaFusionGuard ]
+  //
+  // TODO: we also need to assert/check reduction axes and replace it with
+  // constants in `CudaFusionGroup`
+  Node* typecheck_node = fusion->owningGraph()
+                             ->create(prim::CudaFusionGuard, inputs_to_check, 1)
+                             ->insertBefore(fusion);
+  Value* typecheck_result = typecheck_node->output();
+  typecheck_node->tys_(attr::types, guard_types);
+
+  std::unordered_map<Value*, Value*> typechecked_inputs;
+
+  // Insert if block
+  auto versioning_if =
+      fusion->owningGraph()
+          ->create(prim::If, {typecheck_result}, fusion->outputs().size())
+          ->insertAfter(typecheck_node);
+  for (size_t idx = 0; idx < fusion->outputs().size(); ++idx) {
+    versioning_if->output(idx)->setType(fusion->output(idx)->type());
+    fusion->output(idx)->replaceAllUsesWith(versioning_if->output(idx));
+  }
+  auto true_block = versioning_if->addBlock();
+  auto false_block = versioning_if->addBlock();
+
+  // Fill in the false block. It should contain the unoptimized
+  // copy of the fused subgraph.
+  auto& subgraph = *fusion->g(attr::Subgraph);
+  WithInsertPoint guard(false_block->return_node());
+  const auto subgraph_outputs =
+      insertGraph(*fusion->owningGraph(), subgraph, fusion->inputs());
+  for (Value* output : subgraph_outputs) {
+    false_block->registerOutput(output);
+  }
+
+  // types get copied to the fallback graph, so remove specializations before
+  // replacing
+  // TODO: this is not exposed here, I need to remove that before inserting the
+  //       graph
+  // removeTensorTypeSpecializations(false_block);
+  replaceBlockWithFallbackGraph(false_block, fusion->inputs());
+
+  // Fill in the true block. It has all inputs type-checked and its
+  // body should be the fusion group node.
+  fusion->moveBefore(true_block->return_node());
+  // for (size_t idx = 0; idx < fusion->inputs().size(); ++idx) {
```

Comment:
ditto: why is this commented out?

Reply:
This is an oversight. Thanks for catching this.